### PR TITLE
8314986: Module readability resolution is slow with large numbers of automatic modules

### PR DESCRIPTION
**Pull request by Technici4n**
Fixes the issue (hopefully) by resolving automatic modules and automatic module dependencies after propagation of non-automatic transitive dependencies. The module tests run.

I also added a few asserts to validate that the automatic modules don't read themselves (previously this was the case with > 1 automatic module). Should these asserts be added in more places or is this enough?

Alan also mentioned a conflict with the spec, I'm not sure which spec is being referred to. The documentation of `ResolvedModule#reads` states `A possibly-empty unmodifiable set`, implying that the set can be empty.

Finally, `Configuration#reads` states `// The sets stored in the graph are already immutable sets` but that does not seem to be true. Should something be done about this to limit allocation?

Please let me know!
Cheers

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Integration blockers
&nbsp;⚠️ Too few reviewers with at least role reviewer found (have 0, need at least 1) (failed with the updated jcheck configuration)
&nbsp;⚠️ Whitespace errors (failed with the updated jcheck configuration)

### Issue
 * [JDK-8314986](https://bugs.openjdk.org/browse/JDK-8314986): Module readability resolution is slow with large numbers of automatic modules (**Bug** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15926/head:pull/15926` \
`$ git checkout pull/15926`

Update a local copy of the PR: \
`$ git checkout pull/15926` \
`$ git pull https://git.openjdk.org/jdk.git pull/15926/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15926`

View PR using the GUI difftool: \
`$ git pr show -t 15926`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15926.diff">https://git.openjdk.org/jdk/pull/15926.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15926#issuecomment-1735661422)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/module/Resolver.java [hunk 1]
```diff
@@ -649,6 +606,77 @@ private Map<ResolvedModule, Set<ResolvedModule>> makeGraph(Configuration cf) {
...
+                    assert m2 != null;
+                    m1Reads.add(m2);
+                }
+            }
+
+            // reads all modules in parent configurations
+            for (Configuration parent : parents) {
+                parent.configurations()
+                        .map(Configuration::modules)
+                        .flatMap(Set::stream)
```
Comment (liach) [inline src/java.base/share/classes/java/lang/module/Resolver.java:639]: Can't this and below step be `.forEach(m1Reads::addAll)`?

File: src/java.base/share/classes/java/lang/module/Resolver.java
Comment (liach) [inline src/java.base/share/classes/java/lang/module/Resolver.java:1]: You should update the ending copyright year of this file from 2022 to 2023.

File: src/java.base/share/classes/java/lang/module/Resolver.java [hunk 2]
```diff
@@ -649,6 +606,77 @@ private Map<ResolvedModule, Set<ResolvedModule>> makeGraph(Configuration cf) {
...
+        // in parent configurations
+        for (Map.Entry<ResolvedModule, Set<ResolvedModule>> entry : g1.entrySet()) {
+            ResolvedModule m1 = entry.getKey();
+
+            if (!m1.descriptor().isAutomatic()) {
+                continue;
+            }
+
+            String name = m1.name();
+            Set<ResolvedModule> m1Reads = entry.getValue();
```
Comment (liach) [inline src/java.base/share/classes/java/lang/module/Resolver.java:621]: Why not just this?
```java
m1Reads.addAll(nameToResolved.values());
m1Reads.remove(m1);
```

File: src/java.base/share/classes/java/lang/module/Resolver.java [hunk 3]
```diff
@@ -649,6 +606,77 @@ private Map<ResolvedModule, Set<ResolvedModule>> makeGraph(Configuration cf) {
...
+                    continue;
+                }
+
+                Set<ResolvedModule> m2Reads;
+                if (m2.configuration() == cf) {
+                    // this configuration
+                    m2Reads = g1.get(m2);
+                } else {
+                    // parent configuration, already resolved
+                    // TODO: does this allocate a copy of the set every time?
```
Comment (liach) [inline src/java.base/share/classes/java/lang/module/Resolver.java:665]: It doesn't. `List.copyOf` and `Set.copyOf` returns the passed immutable collection instance if input is already an immutable collection.

**Comment by AlanBateman** [comment]
This issue is on my plate, I will have a PR for this soon. I attached the patch to JDK-8314986 that I was testing in August to skip propagating "requires" through the "requires transitive" edges of the dependence graph. There are several spec issues to work through, the conflict that I mentioned in the JBS issue is that every module is specified to read itself but that is problematic for ResolvedModule::reads to include itself.

File: src/java.base/share/classes/java/lang/module/Resolver.java (hunk 3 above)
Comment (Technici4n) [reply src/java.base/share/classes/java/lang/module/Resolver.java:665]: The input did not seem immutable however

**Comment by Technici4n** [comment]
Hi @AlanBateman, you are still taking care of this issue for jdk 22, right? Else I can update the PR.

**Comment by AlanBateman** [comment]
> Hi @AlanBateman, you are still taking care of this issue for jdk 22, right? Else I can update the PR.

Yes, it's on my list.

**Pull request closed without merging**